consul/resource_consul_autopilot_config.go

```diff
@@ -0,0 +1,150 @@
+package consul
+
+import (
+	"fmt"
+	"time"
+
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+const (
+	consulAutopilotConfigDatacenter              = "datacenter"
+	consulAutopilotConfigCleanupDeadServers      = "cleanup_dead_servers"
+	consulAutopilotConfigLastContactThreshold    = "last_contact_threshold"
+	consulAutopilotConfigMaxTrailingLogs         = "max_trailing_logs"
+	consulAutopilotConfigServerStabilizationTime = "server_stabilization_time"
+	consulAutopilotConfigRedundancyZoneTag       = "redundancy_zone_tag"
+	consulAutopilotConfigDisableUpgradeMigration = "disable_upgrade_migration"
+	consulAutopilotConfigUpgradeVersionTag       = "upgrade_version_tag"
+)
+
+func resourceConsulAutopilotConfig() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceConsulAutopilotConfigCreate,
+		Update: resourceConsulAutopilotConfigUpdate,
+		Read:   resourceConsulAutopilotConfigRead,
+		Delete: resourceConsulAutopilotConfigDelete,
+
+		Schema: map[string]*schema.Schema{
+			consulAutopilotConfigDatacenter: &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			consulAutopilotConfigCleanupDeadServers: &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+			consulAutopilotConfigLastContactThreshold: &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "200ms",
```

Comment:
Do you think we should just omit this and let Consul set the default upstream? That default may change in the future. It might be better to say this is computed if not set.

Reply:
I first tried to do this with `Operator.AutopilotCASConfiguration` but I got issues with defaults which would always differ from the state and that would result in a perpetual diff.

I did not think of it at the time but could I use `DefaultFunc` to make a call to Consul and fetch current settings here? I think it would make one call per attribute unless I find a way to cache the call. This behavior would be more predictable to users and the extra complexity is probably worth it.

Reply:
Hm I think using defaultfunc would basically just be doing `computed: true` which I think makes sense for this. Do you see what I mean? It is a computed attribute meaning when it isn't set it will be computed remotely and then written back.

Reply:
I tried that but it seems that setting `Computed: true` it is then impossible to detect when between:
```
resource "consul_autopilot_config" "config" {}
```
and
```
resource "consul_autopilot_config" "config" {
	upgrade_version_tag =  ""
}
```
 The first should be no-op but the second should remove the current value of `upgrade_version_tag` and set `""` instead. It seems to me that it is not possible to detect whether a value has been defined by the user when it is computed: https://github.com/hashicorp/terraform/blob/3d908f56aa6588b9a2db7cdf6b36bdf7b45d781f/helper/schema/resource_data.go#L123

Maybe a `GetOkExistsComputed` could be used here?

Reply:
Hm from the docs there it seems like that is only intended for booleans. Perhaps we should just hardcode the defaults for now. It isn't ideal but I'm comfortable with it given the constraints as I understand them.